multicluster/controllers/multicluster/node_controller.go

```diff
@@ -0,0 +1,199 @@
+		return ctrl.Result{}, err
+	}
+
+	_, isGWNode := node.Annotations[common.GatewayAnnotation]
+	var err error
+	var gwIP, internalIP string
+	if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
+		klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+		return ctrl.Result{}, nil
+	}
+	gwNamespacedName := types.NamespacedName{
+		Name:      node.Name,
+		Namespace: r.namespace,
+	}
+
+	if err := r.Client.Get(ctx, gwNamespacedName, gwNode); err != nil {
+		if apierrors.IsNotFound(err) && isGWNode {
+			gwNode.GatewayIP = gwIP
+			gwNode.InternalIP = internalIP
+			if err := r.Client.Create(ctx, gwNode, &client.CreateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.installedGWNodes.Add(gwNode)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !isGWNode {
+		if err := r.Client.Delete(ctx, gwNode, &client.DeleteOptions{}); err != nil {
+			return ctrl.Result{}, client.IgnoreNotFound(err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	gwNode.GatewayIP = gwIP
+	gwNode.InternalIP = internalIP
+	if err := r.updateGateway(gwNode); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *NodeReconciler) updateGateway(gwNode *mcsv1alpha1.Gateway) error {
+	cached, exist, _ := r.installedGWNodes.GetByKey(gwNode.Namespace + "/" + gwNode.Name)
+	if exist {
+		cachedGWNode := cached.(*mcsv1alpha1.Gateway)
+		if cachedGWNode.GatewayIP == gwNode.GatewayIP && cachedGWNode.InternalIP == gwNode.InternalIP {
+			klog.V(2).InfoS("No change for the Gateway, skip updating", "gateway", klog.KObj(gwNode))
+			return nil
+		}
+	}
+	if err := r.Client.Update(ctx, gwNode, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	r.installedGWNodes.Update(gwNode)
+	return nil
+}
+
+func (r *NodeReconciler) getGatawayNodeIP(node *corev1.Node) (string, string, error) {
+	var gatewayIP, internalIP string
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			if r.precedence == mcsv1alpha1.PrecedencePrivate {
+				gatewayIP = addr.Address
+			}
+			internalIP = addr.Address
+		}
+		if r.precedence == mcsv1alpha1.PrecedencePublic && addr.Type == corev1.NodeExternalIP {
+			gatewayIP = addr.Address
+		}
+	}
+
+	if ip, ok := node.Annotations[common.GatewayIPAnnotation]; ok {
+		parsedIP := net.ParseIP(ip)
+		if parsedIP == nil {
+			return "", "", fmt.Errorf("the Gateway IP annotation %s on Node %s is not a valid IP address", ip, node.Name)
+		}
+		gatewayIP = ip
+	}
+
+	if gatewayIP == "" || internalIP == "" {
+		return "", "", fmt.Errorf("no valid IP address for Gateway Node %s", node.Name)
+	}
+	return internalIP, gatewayIP, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NodeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Node{}).
```

Comment:
Not sure how often node update events can occur, but it seems that for update events we only care about annotation changes. https://github.com/kubernetes-sigs/controller-runtime/blob/3f265c36d7bf597facc9b0363b48ff21a68cb782/pkg/predicate/predicate.go#L179 seems promising

Reply:
Actually we also need to watch Node's status change which includes its Internal and External IP.

Reply:
Got it, thanks for the explanation